### DISCUSSION TITLE
React Native SDK - Can not login on android platform

### DISCUSSION THREAD

**Discussion by huyta** [Web3Auth/Web3Auth#812] [Help]
Hello,

I am integrating web3auth into my React native app, it works well on iOS

But on Android devices, I got the error: **[Error: The WebBrowser auth session is in an invalid state with no redirect handler when one should be set]**

Do you have any idea about this error?

My codes:

```
const info = await web3auth.login({
                loginProvider: LOGIN_PROVIDER.JWT,
                redirectUrl: resolvedRedirectUrl,
                dappShare: true,
                extraLoginOptions: {
                    verifierIdField: "sub",
                    id_token: idToken,
                    domain: "app-staging",
                },
            });

            console.log(info);

```

Thanks,
Huy

> Note: This works on `"react-native": "0.70.4"`. This has been update by the moderator.

**Comment by shahbaz17** [comment]
Hey @huyta 

Please share your `resolvedRedirectUrl` here.

**Reply by huyta** [reply]
Hi @shahbaz17,

My resolvedRedirectUrl: huytran://auth

The same code base is working on iOS

**Reply by shahbaz17** [reply]
And what's your android package name?

**Reply by huyta** [reply]
Hi @shahbaz17,

my android package name : world.op3n.mobile.development 
I did try both: 
```
 <data android:scheme="world.op3n.mobile.development" />
 <data android:scheme="huytran" />

```

**Reply by shahbaz17** [reply]
Can you try `w3a://world.op3n.mobile.development/auth` as the `resolvedRedirectUrl` for Android? 

Note that `w3a://world.op3n.mobile.development/auth` should also be whitelisted on the Web3Auth Dashboard.

**Reply by huyta** [reply]
Hi @shahbaz17,

Tried it but no luck

I dont think the problem is redirect url since the sample(https://github.com/Web3Auth/examples/tree/main/react-native/bare) is working with my custom verifier and redirect url 

The problem should be on WebBrowser auth session

**Reply by shahbaz17** [reply]
I have reported it to our engineers. Will get back to this once this is resolved.

**Reply by huyta** [reply]
Hello @shahbaz17,

Any update on this ?

FYI:
+ My exist project is based on: "react-native": "0.64.2", (got the error: [Error: The WebBrowser auth session is in an invalid state with no redirect handler when one should be set])
+ I just create new RN project ("react-native": "0.70.4") and integrate web3auth from scratch, then it works OK

Do we need any special config for RN 0.64.2?

Thanks,
Huy

**Reply by huyta** [reply]
Hello @shahbaz17 ,

I can confirm that web3auth login is not working on React Native 0.64.2 (Always got error: The WebBrowser auth session is in an invalid state with no redirect handler when one should be set)

**Reply by shahbaz17** [reply]
Hey @huyta 

We need to test this and this breaks on "react-native": "0.64.2" and works on "react-native": "0.70.4". Then we should add this as a requirement.

**Comment by shahbaz17** [comment]
Works on `"react-native": "0.70.4"`

**Reply by huyta** [reply]
Did you find the problem? Why it does not work on RN 0.64.2?

**Reply by shahbaz17** [reply]
@yashovardhan Can you test this with `RN 0.64.2`?

**Reply by huyta** [reply]
Hello @yashovardhan,

Confirmed that Web3Auth is not working on React Native 0.64.X (Android only, iOS is working ok)

I pushed the test project here https://github.com/huyta/Web3AuthTest so you can have a look at the error

Thanks,
Huy

**Reply by yashovardhan** [reply]
I was checking the app, a dependency was out of date considering you're using `"react-native": "0.65.0"`

make sure `react` & `react-test-renderer` are at  `"17.0.2"`

Anyway while running the app, I did't find any errors as per say, the only thing was the app was not redirecting back, I changed the scheme to be `${scheme}://openlogin` and it was working after that. Please have a look once again.

**Comment by yashovardhan** [comment]
update: this doesn't work properly. Redirect won't give you proper return. However all other things remain for SDK implementation.

I tested this on RN version `0.64.2` -  Here are the things I needed to do:

- had to make sure my minimum SDK compile version in my `build.gradle` is as follows:
```
buildToolsVersion = "31.0.0"
minSdkVersion = 21
compileSdkVersion = 31
targetSdkVersion = 31
```

- Add the intent filter with scheme defined in my `AndroidManifest.xml`

```
<intent-filter>
        <action android:name="android.intent.action.VIEW" />
        <category android:name="android.intent.category.DEFAULT" />
        <category android:name="android.intent.category.BROWSABLE" />
        <data android:scheme="web3authrnexample" />
</intent-filter>
```

- SDK version 31 requires you to explicitly define `android:exported="true"` in `AndroidManifest.xml`
```
<activity
        android:name=".MainActivity"
        android:label="@string/app_name"
        android:configChanges="keyboard|keyboardHidden|orientation|screenSize|uiMode"
        android:launchMode="singleTask"
        android:windowSoftInputMode="adjustResize"
        android:exported="true">
```

Following this, I was able to run it properly.

**Reply by huyta** [reply]
Hi @yashovardhan,

It run properly, but the login always return exception on Android: 

**[Error: The WebBrowser auth session is in an invalid state with no redirect handler when one should be set]**

Did you try login successfully?

**Reply by huyta** [reply]
```
const login = async () => {
    try {
      const web3auth = new Web3Auth(WebBrowser, {
        clientId,
        network: OPENLOGIN_NETWORK.TESTNET, // or other networks
      });

      const info = await web3auth.login({
        redirectUrl: resolvedRedirectUrl,
        loginProvider: LOGIN_PROVIDER.GOOGLE,
      });
      console.log('Logged In:', info);
    } catch (e) {
      console.error(e);
    }
  };
```

web3auth.login always return error (The WebBrowser auth session is in an invalid state with no redirect handler when one should be set) on Android RN 0.64.X

Could you check please @yashovardhan

**Reply by yashovardhan** [reply]
please check my comment on the previous post

**Reply by huyta** [reply]
Hi @yashovardhan,
I forgot that I upgraded to RN 0.65.0 in my Test project (https://github.com/huyta/Web3AuthTest), So it work properly

I just rollback to RN 0.64.0 in my test project (https://github.com/huyta/Web3AuthTest)

Could you try again @yashovardhan

**Reply by yashovardhan** [reply]
I tried - it was redirecting back to the app.

**Reply by yashovardhan** [reply]
https://github.com/yashovardhan/web3auth-rn0.64-test please check this app

**Reply by huyta** [reply]
Hi @yashovardhan,

I tried the sample you provided (https://github.com/yashovardhan/web3auth-rn0.64-test)

I got an error (see image link below) and still not login

https://ibb.co/k2LxfPN

Could you have a look please

**Reply by yashovardhan** [reply]
@shahbaz17 can you please check as well if you're getting this error?

**Reply by huyta** [reply]
Run the sample you provided here https://github.com/yashovardhan/web3auth-rn0.64-test, still got the same error **The WebBrowser auth session is in an invalid state with no redirect handler when one should be set**

**Reply by huyta** [reply]
I guess, the problem should be on @toruslabs/react-native-web-browser on RN 0.64.X  @yashovardhan @shahbaz17

**Reply by yashovardhan** [reply]
I checked around for a bit around this. Basically my app was working because of module catching in my system. 0.64.X is not supported by us.

**Comment by yashovardhan** [comment]
React Native version 0.64.X is not supported by Web3Auth SDK. This is because we need to use the ` @toruslabs/react-native-web-browser` which is a port of `expo-web-browser` module for the bare flow. This module port in the latest change doesn't support React Native below v0.65.

 
Hence, Web3Auth React Native SDK supports React Native v0.65 and above.

Thanks for bringing this to our attention @huyta
